### PR TITLE
Support for NTLM authentication added

### PR DESCRIPTION
**Pull request by m7913d**
To support NTLM authentication, a database is added as an authentication source. Currently, only the configuration file is supported as a database. Database authentication supports Basic and NTLM authentication protcols.

ServerConfig.BasicAuthEnabled renamed to LocalEnabled as Basic auth can be used with Database or Local.

Advantage of this new database authentication option is that it supports the default windows client without the need to setup OpenID Connect or Kerberos.

### CODE REVIEW

**Review by bolkedebruin — CHANGES_REQUESTED** [review]
First of all thanks for the extension! I have been on the hunt for a server side NTLM implementation that has been proven and provides a reasonable degree of security. I did not come across this one yet so I am happy you found it.

### Architecture
A couple of notes on the architecture. As `rdpgw` could be directly internet facing the `rdpgw` process should not be able to access any database of passwords either in memory or on disk. This is why `rdpgw-auth` is there. `rdpgw-auth` can run as root (or different user) so it separates the concerns. Hence, I think this should be there. 

I think approaching this as 'database' authentication isn't fitting. This clutters the code and I don't think the audience of `rdpgw` would immediately grasp what `database` authentication means.  You can't google "microsoft remote desktop gateway database authentication"

So, I would like to see this first as NTLM authentication. Basic authentication through a flat file can then be a extension of `rdpgw-auth` if required opaque to `rdpgw`.

### License and maintainability
I see the NTLM library was forked from ThomsonReuters with a BSD 4-Clause License. This contains an advertising clause which would need to be incorporated.

You have forked the NTLM library to your own account. Given all the abandoned repos before this fork, what is your intention in maintining it? 

### Testing
Please add tests and remove debugging information that could conmtain sensitive information (NTLM hashes)

**Comment by m7913d** [comment]
Thanks for your review.

### `database` option

Regarding the name `database` authentication, I think most users are interested in the source used to validate the credentials, not the protocol (e.g. Basic or NTLM) used to communicate between the client and the gateway. In fact, most users would expect the gateway to work with any compliant RDP client (especially the standard Windows client), so not supporting `NTLM` could be considered a bug in this regard. For consistency, I think the user should choose either the source (`local` or `database`) or the protocol (`Basic` or `NTLM`). Nevertheless, if you want to rename `database` to `NTLM`, that's fine by me.

### Architecture
So, you want to move the NTLM validation to `rdpgw-auth`? Or to a new helper program, e.g. `rdpgw-ntlm`?

Which part of the code do you want to move to this helper program?
 - All NTLM message parsing and generation (Negotiate, Challenge and Authenticate message). Disadvantage is that you may consider now the helper program as directly internet facing too.
 - Retreive the password hash (i.e. ntowfv1/2 and ntowfv1/2) from the helper program. Disadvantage is that the password hash is a password equivalent, gaining little improved security. Only advantage is that the password hash is in the memory of the main process for a shorter period of time
 - Only pass the Authenticate message and Server challenge to the helper program. This is the securest option, but it requires the most modifications and/or insight into the go-ntlm library internals. The authenticate message (or at least a substantial part of it) received from the client, is still passed to the helper program, as it is needed to perform the authentication. This may be a possible attack vector too.
 
Which threat model do you take into account?

### License
Where do you want to include it?

### Maintainability
Ideally, I would like to push my changes to an upstream. The staaldraad repo (from which I pulled the source code) or the Psiphon project (from which I included some bug fixes) might be better suited as maintainers, but I don't know yet if they are interested.

If you want to be the maintainer yourself, that's fine by me too.

### Testing
I will have a look at it.

**Comment by bolkedebruin** [comment]
On "database", I think we differ of opinion here :-). We can (eventually) enable all authentication mechanisms at the same time and they are all backed up by a database some way. So let's set this as ntlm please. 

Also because I'd like to be able to have basic and ntlm authentication at the same time "rdpgw-auth" is the right place.

I consider the "database" that needs to be read as sensitive as /etc/shadow. Hence accessing it in a separate process with a separate uid. The rpc protocol can be extended to support the challenge response mechanism. Essentially relaying between the client and rdpgw-auth. I guess this is option three but I don't see why this would require knowledge of the go-ntlm internals. 

For the license I think putting it in licenses naming it "thomson-reuters" and including the line "this software..." When running rdpgw-auth without arguments or with -h should be okay.  Open to suggestions here of course. Just kind of making this up

**Comment by m7913d** [comment]
### Architecture

If you split the generation of the Challenge and the validation of the Authenticate message to a different process process (=option 3), we need to pass at least the Server challenge (which I consider a go-ntlm internal as it is currently not exposed in the interface) and maybe other things (such as the negotiated flags) to `rdpgw-auth`.

The go-ntlm library rather assumes that a single `ServerSession` instance is used to generate the Challenge and validate the Authenticate message. See https://github.com/m7913d/go-ntlm?tab=readme-ov-file#sample-usage-as-ntlm-server

However, if you say "The rpc protocol can be extended to support the challenge response mechanism", it gives me the intention that you want to generate the Challenge by `rdpgw-auth`, which is option 1.

**Comment by bolkedebruin** [comment]
> ### Architecture

> However, if you say "The rpc protocol can be extended to support the challenge response mechanism", it gives me the intention that you want to generate the Challenge by `rdpgw-auth`, which is option 1.

Yep :-)

File: cmd/rdpgw/config/configuration.go [hunk 1]
```diff
@@ -221,9 +228,13 @@ func Load(configFile string) Configuration {
 		log.Fatalf("host selection is set to `signed` but `querytokensigningkey` is not set")
 	}
 
-	if Conf.Server.BasicAuthEnabled() && Conf.Server.Tls == "disable" {
+	if Conf.Server.LocalEnabled() && Conf.Server.Tls == "disable" {
 		log.Fatalf("basicauth=local and tls=disable are mutually exclusive")
 	}
+        
+        if Conf.Server.DatabaseEnabled() && Conf.Server.KerberosEnabled() {
```
[the remark this answers is not in the document]
Comment (m7913d) [reply cmd/rdpgw/config/configuration.go]: Feel free to add a commit to allow stacking Kerberos and NTLM. As I currently do not have a Kerberos setup at hand, I'm reluctant to add support for it without being able to test whether it is working as expected.

File: cmd/rdpgw/web/ntlm.go [hunk 2]
```diff
@@ -0,0 +1,222 @@
...
+}
+
+func (h *NTLMAuthHandler) NTLMAuth(next http.HandlerFunc) http.HandlerFunc {
+        return func(w http.ResponseWriter, r *http.Request) {
+                //log.Printf("NTLM request")
+
+                authPayload, authMode, err := h.getAuthPayload(r)
+                if err != nil {
+                        log.Printf("Failed parsing auth header: %s", err)
+                        w.Header().Add("WWW-Authenticate", `NTLM,Basic realm="restricted", charset="UTF-8"`)
```
[the remark this answers is not in the document]
Comment (m7913d) [reply cmd/rdpgw/web/ntlm.go]: NTLM may be not an IANA standard, but it is documented by Microsoft: https://learn.microsoft.com/en-us/dotnet/framework/wcf/feature-details/understanding-http-authentication

The ` WWW-Authenticate: NTLM` is also observed by others: https://davenport.sourceforge.net/ntlm.html#ntlmHttpAuthentication

I also observed this scheme being used by Remmina (RDC for POSIX).

File: cmd/rdpgw/web/ntlm.go [hunk 3]
```diff
@@ -0,0 +1,222 @@
...
+                        id.SetAuthenticated(true)
+                        id.SetAuthTime(time.Now())
+                        next.ServeHTTP(w, identity.AddToRequestCtx(id, r))
+                }
+        }
+}
+
+func (h *NTLMAuthHandler) getAuthPayload (r *http.Request) (payload string, authMode ntlmAuthMode, err error) {
+        authorisationEncoded := r.Header.Get("Authorization")
+        if authorisationEncoded[0:5] == "NTLM " {
```
[the remark this answers is not in the document]
Comment (m7913d) [reply cmd/rdpgw/web/ntlm.go]: At least Remmina does send this header.

**Comment by m7913d** [comment]
I refactored the code as requested, except for:
 - NTLM header is "valid"/used by some clients (see comments above)
 - Stacking Kerberos and NTLM not implemented (see above)

**Comment by m7913d** [comment]
Concerning the maintainability of the go-ntlm library, I contacted some projects using the library (psiphon and sematext) and recent contributors (staaldraad), but only staaldraad responded so far.

I'm fine with reviewing PR to the go-ntlm library, but I will not actively contribute to the further development. It may be better to host it under your own account. 

As NTLM is considered rather an old/being phased out authentication mechanism, I don't expect many new features will be added to the library, but I think supporting the NTLM protocol is still very relevant for the rdpgw project to support the default Microsoft remote desktop client.

File: shared/auth/auth.pb.go [hunk 4]
```diff
@@ -1,7 +1,7 @@
 // Code generated by protoc-gen-go. DO NOT EDIT.
 // versions:
-// 	protoc-gen-go v1.28.1
-// 	protoc        v3.21.5
+// 	protoc-gen-go v1.25.0-devel
```
Comment (bolkedebruin) [inline shared/auth/auth.pb.go:3]: I guess it works, but your stack looks a bit outdated :-)

**Comment by bolkedebruin** [comment]
Awesome! I probably will move away from storing user credentials into the config file. I'd rather have something like /etc/smbpasswd that is encrypted with a key from the environment. But for now looks good.

File: shared/auth/auth.pb.go (hunk 4 above)
Comment (m7913d) [reply shared/auth/auth.pb.go:3]: Sorry, I just used the default generator installed on my system to generate the files. Feel free to regenerate the files with more recent versions.

**Comment by m7913d** [comment]
> Awesome! I probably will move away from storing user credentials into the config file. I'd rather have something like /etc/smbpasswd that is encrypted with a key from the environment. But for now looks good.

Encrypting using a reversible algorithm would be possible. We can also store the NTLM hash of the user credentials (i.e. [NTOWFv2](https://learn.microsoft.com/en-us/openspecs/windows_protocols/ms-nlmp/5e550938-91d4-459f-b67d-75d70009e3f3), see [go-ntlm](https://github.com/m7913d/go-ntlm/blob/3aa26fa3cfeb079b6fe16fa27a6439d1a278e01b/ntlm/ntlmv2.go#L517)), instead of the plain password. Just take into account that this hash is a password equivalent, i.e. knowing this hash is enough to authenticate, but it still prevents leaking the original password in case the user reused it for another application.

**Comment by m7913d** [comment]
Thanks for reviewing/merging!

**Comment by m7913d** [comment]
@ryanblenis I'm not very familiar with LDAP. If you could retrieve the NT and LM hash from LDAP, then it should be relatively easy to to support NTLM auth with LDAP as backend.

Stacking authentication protocols only means that the `rdpgw` server indicates to the client that it supports both authentication protocols, but I think most (or all) clients will not fall back to Basic if the server indicates that it also supports NTLM, as the client probably assumes that if the password is rejected using NTLM, it will also be rejected using Basic auth. I think it's somewhat uncommon to use different password backends depending on the authentication protocol used with the client, but this is partly the result of the authentication protocols (between client and gateway server) not being compatible with all password backends (i.e. for local auth, the client needs to pass the plain password to the gateway server, which NTLM doesn't do).

So, to make stacking work, you should:
 - configure the same passwords in the different password backends (used by the different auth protocols)
 - or explicitly tell the client which protocol to use (if supported by the client)
 - or ensure the same password backend is used for the stacked authentication protocols (probably requires changes to the library). In my first implementation, I supported validating the password against the same password database (i.e. currently the config file) for both NTLM and Basic auth, but due to the partial rework requested by bolkedebruin, this feature was lost, but it may be added again.

Note that I don't use stacking at the moment.

_Disclaimer_:  I'm not the maintainer of this library. So, information may be inaccurate.

**Comment by ryanblenis** [comment]
@m7913d Thanks for the quick response. I don't believe you can retrieve the hashes from LDAP, but I believe you can bind with LDAP via hash (at least this function would have me believe https://github.com/go-ldap/ldap/blob/master/bind.go#L465 ), I just can't seem to find out where the appropriate hash is accessible for this purpose.

**Comment by m7913d** [comment]
@ryanblenis  I guess that this it the NT or LM password hash as stored in the LDAP database as I don't see any challenge/response mechanism present. If this is indeed the case, it is not possible to use this function as the client hashes this password has again with a randomly generated challenge.

**Pull request merged**